src/analyses/base.ml

```diff
@@ -1597,14 +1635,67 @@ struct
         if M.tracing then M.tracel "inv" "Unhandled operator %a\n" d_binop op;
         a, b
     in
+    let inv_bin_float (a, b) fkind c op =
+      let meet_bin a' b'  = FD.meet a a', FD.meet b b' in
+      let meet_com oi = (* commutative *)
+        meet_bin (oi c b) (oi c a) in
+      let meet_non oi oo = (* non-commutative *)
+        meet_bin (oi c b) (oo a c) in
+      match op with
+      | PlusA  -> meet_com FD.sub
+      | Mult   ->
+        (* refine x by information about y, using x * y == c *)
+        let refine_by x y = if FD.is_exact y then FD.meet x (FD.div c y) else x
```

Comment:
When doing this refinement and others, have you considered which IEEE 754 floating point identities are actually guaranteed to hold? For example, their addition isn't even associative, so it could be quite likely that the identities used for these refinements aren't correct either. Especially since right now the analysis internally uses the same floating point numbers.

Reply:
This was indeed a major unsoundness. This should now be fixed. The idea behind it is an adaption of the proposed method [here](https://hal.inria.fr/inria-00540299/document).

Reply:
Wow, that's some impressive work! You should put that reference link into a comment before this code, such that it's easier to later find what it's based on without having to dig into the comments of this PR.

Also, did any of the existing tests or SV-COMP show this unsoundness? If not, but you had some tests yourself to validate the soundness of the new implementation compared to the old one, it'd be very valuable to have them as tests. Then we could even submit them to SV-COMP.

Reply:
The sv-comp benchmark didn't show this unsoundness in the last run. However, this could also be due to a crash that frequently happened in similar situations.
Nonetheless i have added some more dedicated tests for the respective border cases in regression test `57/11`.